enterprise/backend/src/metabase_enterprise/ai_sql_fixer/api.clj

```diff
@@ -1,13 +1,122 @@
 (ns metabase-enterprise.ai-sql-fixer.api
   "`/api/ee/ai-sql-fixer/` routes"
   (:require
+   [clojure.set :as set]
    [metabase-enterprise.metabot-v3.client :as metabot-v3.client]
    [metabase.api.macros :as api.macros]
    [metabase.api.routes.common :refer [+auth]]
    [metabase.driver.util :as driver.u]
-   [metabase.models.persisted-info :as persisted-info]
+   [metabase.query-analysis.native-query-analyzer :as nqa]
    [metabase.query-processor.middleware.permissions :as qp.perms]
-   [metabase.util.malli.schema :as ms]))
+   [metabase.util :as u]
+   [metabase.util.malli.schema :as ms]
+   [toucan2.core :as t2]
+   [toucan2.realize :as t2.realize])
+  (:import
+   (java.io StringWriter Writer)
+   (org.apache.commons.text.similarity LevenshteinDistance)))
+
+(set! *warn-on-reflection* true)
+
+;; some arbitrary limits
+(def ^:private max-schema-sample-tables
+  "If the number of tables in the database doesn't exceed this number, we send them all to the agent."
+  10)
+
+(def ^:private max-schema-candidate-tables
+  "The maximum number of tables we process to find the ones that might be relevant for the SQL query to fix."
+  10000)
+
+(def ^:private max-distance
+  "The maximal edit distance at which we consider two table (or schema) names similar."
+  4)
+
+(def ^:private ^LevenshteinDistance matcher (LevenshteinDistance. (int max-distance)))
+
+(defn- similar?
+  [left right]
+  (nat-int? (.apply matcher (u/lower-case-en left) (u/lower-case-en right))))
+
+(defn- matching-tables?
+  [{tschema :schema, tname :name} {uschema :schema, uname :name} {:keys [match-schema?]}]
+  (and (similar? uname tname)
+       (or (not match-schema?)
+           (nil? uschema)
+           (nil? tschema)
+           (similar? uschema tschema))))
+
+(defn- find-matching-tables
+  [database-id unrecognized-tables used-ids]
+  (into []
+        (keep (fn [table]
+                (when (some #(matching-tables? table % {:match-schema? false}) unrecognized-tables)
```

Comment:
I've decided not to match on the schema, because I hope that having lots of schemas with the same table name is not a frequent use case. I guess it might used for distinguishing tenants, staging and prod environments, etc. The worst case is probably if regular copies of the schema are created.

Reply:
With dbt this can definitely happen. However, from my experience, usually folks would hide those other schemas in Metabase. Is this an information we have available and could use (i.e. not consider hidden tables / schemas)?
Or are those automatically excluded?

Reply:
I have to check this. The intention is not to look at hidden stuff.

Reply:
I suppose schemas that are unused copied should be excluded from synching altogether, but I added checks to ignore inactive tables and tables with nonstandard visibility.